File: server-shared/src/main/java/eu/darken/porter/core/CallerIdentity.java
package eu.darken.porter.core;

import android.os.Binder;

import rikka.shizuku.server.util.UserHandleCompat;

/**
 * Who made a call. Captured once at the endpoint the call arrived on and handed down from there, so
 * that nothing below the endpoint has to ask Binder again.
 */
public final class CallerIdentity {

    public final int uid;
    public final int pid;

    public CallerIdentity(int uid, int pid) {
        this.uid = uid;
        this.pid = pid;
    }

    public static CallerIdentity fromBinder() {
        return new CallerIdentity(Binder.getCallingUid(), Binder.getCallingPid());
    }

    public int userId() {
        return UserHandleCompat.getUserId(uid);
    }

    public int appId() {
        return UserHandleCompat.getAppId(uid);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof CallerIdentity)) return false;
        CallerIdentity other = (CallerIdentity) obj;
        return uid == other.uid && pid == other.pid;
    }

    @Override
    public int hashCode() {
        return 31 * uid + pid;
    }

    @Override
    public String toString() {
        return "CallerIdentity{uid=" + uid + ", pid=" + pid + "}";
    }
}

File: server-shared/src/main/java/eu/darken/porter/core/CallerIdentity.kt
package eu.darken.porter.core

import android.os.Binder
import rikka.shizuku.server.util.UserHandleCompat

/**
 * Who made a call. Captured once at the endpoint the call arrived on and handed down from there, so
 * that nothing below the endpoint has to ask Binder again.
 */
class CallerIdentity(val uid: Int, val pid: Int) {

    fun userId(): Int = UserHandleCompat.getUserId(uid)

    fun appId(): Int = UserHandleCompat.getAppId(uid)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is CallerIdentity) return false
        return uid == other.uid && pid == other.pid
    }

    override fun hashCode(): Int = 31 * uid + pid

    override fun toString(): String = "CallerIdentity{uid=$uid, pid=$pid}"

    companion object {
        fun fromBinder(): CallerIdentity = CallerIdentity(Binder.getCallingUid(), Binder.getCallingPid())
    }
}
